multiversion: avoid untrodden paths during upgrade

matklad opened the pull request:
During upgrades, tigerbeetle dynamically re-executed itself at a different version. There can be only one tigerbeetle instance running at a time, due to any of:
- data file lock,
- incoming port,
- available RAM. On POSIX, execve semantics of replacing the parent process gives us this for free. On Windows, parent and child cooperate, with the child blocking until the parent exit. Blocking is achieved by passing down an anonymous pipe, a-la <https://matklad.github.io/2023/10/11/unix-structured-concurrency.html>

The previous implementation attempted to achieve a similar effect by de-initing the replica in the data file but that's actually panic prone (storage asserts, on deinit, that there are no next ticks scheduled), and, as a general rule, shutdown code that runs only on one operating system only during upgrade is bound to become buggy sooner or later.

matklad commented:
See manual testing log https://github.com/tigerbeetle/tigerbeetle/commit/fc936a56ff2a8ecf8551bd6a0d12a03a531ce573

I tried to get "in-place upgrades" integration test going, but that hit assert(absPath(argv[0])) assertion which I think is pre-existing, and which I don't heave enough of breath to fix in this dive!

cb22 commented on src/multiversioning.zig:
```diff
@@ -1305,23 +1305,55 @@ pub const Multiversion = struct {
                     null,
                     null,
-                    std.os.windows.FALSE,
+                    std.os.windows.TRUE, // bInheritHandles
```
Hmmmmm!

What about either using a randomly generated named pipe, and passing that in the env, or using `PROC_THREAD_ATTRIBUTE_HANDLE_LIST` to limit the inherited handles?

matklad replied:
🤔 I think named pipe won't work, as it won't get the desired "autoclose when the parent exists" behavior. I'd also prefer this to not be visible outside

So `PROC_THREAD_ATTRIBUTE_HANDLE_LIST` is _probably_ the way to go, though I need to figure out how to use that from Zig, as I think it isn't wrapped...

Actually, I think I don't even need a pipe here? I can pass handle of the parent process to the child process, I _think_ windows handles are safe (they are handles, not pids). Though I still need to inherit it properly, which is the annoying part!

matklad replied:
Ok, actually there's a good way to do this: if I use a named mutex, I avoid both the need to inherit the handles, _and_ I get the correct semantics of resource being released when the parent process exit. 

I implemented that .... and I am getting AccessDenied when opening the data file!

matklad replied:
My next hypothesis is that I actually have to wait until the parent process exits, rather than relying on the fact that the mutex is unlocked _after_ file lock being released. 

so I think my next step is to create a named pipe, and send parents own handle over this pipe, so that the child can wait for it to exit

cb22 commented on src/multiversioning.zig:
```diff
@@ -1386,6 +1418,71 @@ pub fn self_exe_path(allocator: std.mem.Allocator) ![:0]const u8 {
+
+pub fn pipe_parse(buffer: []const u16) !std.os.windows.HANDLE {
+    var utf8_buffer: [31]u8 = undefined;
+    const ut8_size = try std.unicode.utf16LeToUtf8(&utf8_buffer, buffer);
```
Typo? (ut8)

cb22 commented on src/multiversioning.zig:
```diff
@@ -1386,6 +1418,71 @@ pub fn self_exe_path(allocator: std.mem.Allocator) ![:0]const u8 {
+    try T.check(std.math.maxInt(usize));
+}
+
+// During upgrades, tigerbeetle dynamically re-executed itself at a different version.
```
```suggestion
// During upgrades, tigerbeetle dynamically re-executes itself at a different version.
```

matklad committed:
tidy: correctly detect `pub extern "libname" fn`

matklad committed:
multiversion: avoid untrodden paths during upgrade

During upgrades, tigerbeetle dynamically re-executes itself at a different version.
There can be only one tigerbeetle instance running at a time, due to any of:
- data file lock,
- incoming port,
- available RAM.
On POSIX, execve semantics of replacing the parent process gives us this for free.
On Windows, parent and child cooperate, with the child blocking until the parent exit.
Specifically:
1. Parent creates a named pipe with a random name.
2. Parent passes the name of this pipe to the child via inherited env variable.
3. Parent duplicates its own handle into the child process.
4. Parent passes the value of this handle to the child via anonymous pipe
   (We don't want to pass parent handle _directly_ to avoid handle inheritance)
5. Child gets the name of the pipe trough env.
6. Child receives the parent's handle through the pipe.
7. Child waits for parent to exit.

The previous implementation attempted to achieve a similar effect by
de-initing the replica in the data file but that's actually panic prone
(storage asserts, on deinit, that there are no next ticks scheduled),
and, as a general rule, shutdown code that runs only on one operating
system only during upgrade is bound to become buggy sooner or later.

matklad commented:
Should be worth another look now @cb22!

matklad commented:
Not enabling windows tests just yet, as they fail at

```
        const path = arg_iterator.next().?;
        assert(std.fs.path.isAbsolute(path));
```

which I think is an orthogonal issue.

cb22 commented on src/multiversioning.zig:
```diff
@@ -1306,6 +1306,36 @@ pub const Multiversion = struct {
                 posix.close(self.target_fd);
 
+                const pipe_name: [*:0]const u16 =
+                    std.unicode.utf8ToUtf16LeStringLiteral("\\\\.\\pipe\\") ++ random_wstr();
```
I've gotten so used to seeing `++` only in the context of fully comptime known values, that this made me stop and stare for a little!

matklad replied:
Yeah, it's excessively neat that `++` requires only `len` to be `comptime`!

cb22 commented on src/tigerbeetle/main.zig:
```diff
@@ -614,22 +616,6 @@ fn replica_release_execute(replica: *Replica, release: vsr.Release) noreturn {
         @panic("release not available");
     }
 
-    if (builtin.os.tag == .windows) {
```
:tada:
